core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -338,6 +364,20 @@ class half_t {
     return lhs;
   }
 
+  template <class T>
```

Comment:
Are these templates instead of overloads because we want to ensure exact type matches (as opposed to things convertible to `float` and `double`)?

Reply:
@nliber, thanks for your review!

I made these templated b/c:
1. yes, I want to ensure exact types matches such that we only upcast for float and double but still downcast to half_t for integral types.
2. to avoid duplicate code for multiple scalar types -- I wanted to reuse the small operator bodies and make it very easy to add upcasting support for other types going forward.

I'm not aware of any difference in behavior between using templates versus overloads here. Can you elaborate?